### PR TITLE
[PIR-Auto-Parallel]refactor recompute pass in PIR mode

### PR DESCRIPTION
**Pull request by waliwali777**
### PR Category

Auto Parallel

### PR Types

Performance

### Description

基于`PIR` 对重计算 pass 进行重构
（参考旧 `IR` 下重计算实现：https://github.com/PaddlePaddle/Paddle/pull/38920 ）
PCard-88114

### CODE REVIEW

File: test/auto_parallel/pir/auto_parallel_recompute_pir_pass_unittest.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+                    segment_num = max(segment_num, idx)
+                    rc_op_num += 1
+                elif op.has_attr("bwd_recompute_id"):
+                    idx = op.attrs()["bwd_recompute_id"]
+                    segment_num = max(segment_num, idx)
+                    rc_op_num += 1
+        return segment_num + 1, rc_op_num
+
+    def run_test_cases(self):
+        self.strategy._recompute.enable = False
```
Comment (liym27) [inline test/auto_parallel/pir/auto_parallel_recompute_pir_pass_unittest.py:189]: why `strategy.recompute` used in pass but `strategy._recompute` used here ?

File: test/auto_parallel/pir/auto_parallel_recompute_pir_pass_unittest.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        base_segment_num, base_rc_op_num = self.get_recompute_message(base_prog)
+        segment_num_1, rc_op_num_1 = self.get_recompute_message(prog_1)
+        segment_num_2, rc_op_num_2 = self.get_recompute_message(prog_2)
+        segment_num_3, rc_op_num_3 = self.get_recompute_message(prog_3)
+
+        assert base_segment_num == 0 and base_rc_op_num == 0
+        assert (
+            base_segment_num < segment_num_1
+            and segment_num_1 < segment_num_2
+            and segment_num_2 < segment_num_3
```
Comment (liym27) [inline test/auto_parallel/pir/auto_parallel_recompute_pir_pass_unittest.py]: Please check results more accurate.
eg.
`assert base_segment_num == XX`
`assert op_num == XX`

File: python/paddle/distributed/auto_parallel/interface.py [hunk 3]
```diff
@@ -228,6 +231,13 @@ def __call__(self, *args, **kwargs):
                 else:
                     output = self._op(*args, **kwargs)
 
+            if paddle.framework.in_pir_mode():
+                block = paddle.static.default_main_program().global_block()
+                rc_end_id = len(block.ops)
+                for idx in range(rc_begin_id, rc_end_id):
+                    rc_op = block.ops[idx]
+                    rc_op.set_int_attr("fwd_recompute_id", _g_recompute_idx)
```
Comment (liym27) [inline python/paddle/distributed/auto_parallel/interface.py:239]: Why use the prefix "fwd" ?

File: test/auto_parallel/pir/auto_parallel_recompute_pir_pass_unittest.py (hunk 2 above)
Comment (waliwali777) [reply test/auto_parallel/pir/auto_parallel_recompute_pir_pass_unittest.py]: done

File: test/auto_parallel/pir/auto_parallel_recompute_pir_pass_unittest.py (hunk 1 above)
Comment (waliwali777) [reply test/auto_parallel/pir/auto_parallel_recompute_pir_pass_unittest.py:189]: `strategy.recompute` (in pass) is the `recompute` attribute of class `paddle.distributed.Strategy()`
`strategy._recompute`(in engine) is the `_recompute` attribute of class `paddle.distributed.fleet.auto.Strategy()`
The two classes all describe the configuration information of `recompute` and can be converted in `to_static` funxtion

File: python/paddle/distributed/auto_parallel/interface.py (hunk 3 above)
Comment (waliwali777) [reply python/paddle/distributed/auto_parallel/interface.py:239]: `fwd_compute_id` indicates the checkpoint information in the forward that needs to be recomputed.
`bwd_recompute_id` corresponds to `fwd_recompute_id` and is newly added in the backward for recompute, which facilitates debugging.
There are comments in the code to provide explanations.

**Review by XieYunshen — APPROVED** [review]

**Review by liym27 — APPROVED** [review]
LGTM

**Pull request merged**